Реализация требований ФЗ спринта 9

loginov-dv opened the pull request.

loginov-dv committed:
refactor(): изменить название пакета taskManagement на managers

loginov-dv committed:
feat(HttpTaskServer): добавить новый класс HttpTaskServer

loginov-dv committed:
feat(HttpTaskServer): добавить поля класса

loginov-dv committed:
feat(TasksHandler): добавить класс-обработчик пути /tasks и реализовать обработку GET /tasks

loginov-dv committed:
feat(TasksHandler): реализовать обработку пути GET /tasks/{id}

loginov-dv committed:
feat(TasksHandler): реализовать обработку пути DELETE /tasks/{id}

loginov-dv committed:
feat(TasksHandler): реализовать обработку пути POST /tasks

loginov-dv committed:
feat(TasksHandler): добавить try-catch для обработки ошибок

loginov-dv committed:
feat(): изменить способ передачи TaskManager из HttpTaskServer в обработчик TasksHandler

loginov-dv committed:
feat(TasksHandler): добавить TypeAdapter для преобразования Duration в long

loginov-dv committed:
feat(TasksHandler): добавить TypeAdapter для преобразования LocalDateTime в String

loginov-dv committed:
feat(TasksHandler): добавить регистрацию TypeAdapter

loginov-dv committed:
refactor(BaseHttpHandler): добавить базовый класс BaseHttpHandler и перенести в него общие для всех обработчиков методы

loginov-dv committed:
feat(BaseHttpHandler): добавить заголовок Content-Type в ответ

loginov-dv committed:
refactor(TasksHandler): рефакторинг

loginov-dv committed:
feat(HttpTaskServer): добавить методы start() и stop()

loginov-dv committed:
refactor(BaseHttpHandler): добавить комментарии

loginov-dv committed:
refactor(CsvUtils): переименовать CSVUtils в CsvUtils

loginov-dv committed:
feat(BaseHttpHandler): добавить константы для HTTP-методов

loginov-dv committed:
refactor(TasksHandler): рефакторинг, улучшение читаемости кода

loginov-dv committed:
fix(): добавить в реализации трекера изменение globalID при добавлении задач

Исправление бага, при котором задачи с id, сгенерированным трекером, не добавлялись, если ранее были добавлены задачи с явным указанием id

loginov-dv committed:
feat(SubtasksHandler): добавить обработчик для пути /subtasks

loginov-dv committed:
feat(EpicsHandler): добавить обработчик для пути /epics

loginov-dv committed:
feat(HistoryHandler): добавить обработчик для пути /history

loginov-dv committed:
feat(PriorityHandler): добавить обработчик для пути /prioritized

loginov-dv committed:
feat(): перенести метод getPrioritizedTasks() в интерфейс TaskManager

loginov-dv committed:
refactor(): устранить дублирование, вынести общие члены в базовый класс

loginov-dv committed:
feat(HttpTaskServer): добавить эндпоинты

loginov-dv committed:
feat(TasksHandler): добавить обработку исключения, возникающего при парсинге даты

loginov-dv committed:
feat(SubtasksHandler): добавить обработку исключения, возникающего при парсинге даты

loginov-dv committed:
feat(BaseHttpHandler): добавить адаптер для TaskStatus и устранить баг с записью даты в LocalDateTimeAdapter

loginov-dv committed:
feat(BaseHttpHandler): изменить записываемое значение в адаптерах с пустой строки на null

loginov-dv committed:
test(HttpTaskServerTasksTest): добавить тесты для пути /tasks

loginov-dv committed:
feat(TasksHandler): доработать логику определения того, что передана новая задача

loginov-dv committed:
fix(HttpTaskServer): убрать тестовые данные

loginov-dv committed:
test(BaseHttpTaskServerTest): создать базовый класс для тестов и перенести в него общие члены

loginov-dv committed:
test(HttpTaskServerSubtasksTest): добавить тесты для пути /subtasks

loginov-dv committed:
fix(SubtasksHandler): добавить логику определения того, что передана новая задача

loginov-dv committed:
test(HttpTaskServerTasksTest): добавить тест на пересечение задач

loginov-dv committed:
test(BaseHttpTaskServerTest): добавить вспомогательный класс для десериализации списков подзадач

loginov-dv committed:
test(HttpTaskServerEpicsTest): добавить тесты для пути /epics

loginov-dv committed:
fix(EpicsHandler): доработать логику определения того, что передана новая задача

loginov-dv committed:
fix(HttpTaskServerSubtasksTest): исправить неточности в тестах

loginov-dv committed:
refactor(HttpTaskServerTasksTest): исправить комментарии

loginov-dv committed:
test(BaseHttpTaskServerTest): добавить вспомогательный класс для десериализации списков эпиков

loginov-dv committed:
feat(TaskNotFoundException): добавить исключение, возникающее при отсутствии в трекере искомой задачи

loginov-dv committed:
feat(): добавить выбрасывание исключения в методах получения задач, изменить тип возвращаемого значения с Optional на конкретный тип

loginov-dv committed:
feat(): добавить обработку исключения TaskNotFoundException в обработчиках эндпоинтов

loginov-dv committed:
test(): добавить в тесты обработку исключения TaskNotFoundException

loginov-dv committed:
refactor(HttpTaskServer): добавить комментарии, убрать неиспользуемые импорты

loginov-dv committed:
refactor(BaseHttpHandler): добавить комментарии и константу для индекса поля id

loginov-dv committed:
refactor(TasksHandler): рефакторинг

loginov-dv committed:
refactor(SubtasksHandler): рефакторинг

loginov-dv committed:
refactor(EpicsHandler): рефакторинг

loginov-dv committed:
refactor(): удалить неиспользуемые импорты

loginov-dv committed:
feat(): добавить в классы задач защищённое служебное поле type для указания типа задачи

loginov-dv committed:
feat(BaseHttpHandler): добавить полиморфный десериализатор для списка задач

loginov-dv committed:
feat(BaseHttpTaskServerTest): зарегистрировать полиморфный десериализатор для списка задач

loginov-dv committed:
test(HttpTaskServerHistoryTest): добавить тесты для пути /history

loginov-dv committed:
test(): изменить название метода, который наполняет трекер тестовыми данными

loginov-dv committed:
test(HttpTaskServerPriorityTest): добвить тесты для пути /prioritized

loginov-dv committed:
refactor(HttpTaskServerTasksTest): создать константы путей, методов, добавить комментарии

loginov-dv committed:
refactor(HttpTaskServerTasksTest): выполнить рефакторинг, добавить рандомизацию в тесты

loginov-dv committed:
refactor(): выполнить рефакторинг, добавить рандомизацию в тесты

loginov-dv committed:
refactor(HttpTaskServerPriorityTest): заменить путь-строку на константу

loginov-dv committed:
refactor(BaseHttpTaskServerTest): удалить неиспользуемые TypeToken

loginov-dv committed:
refactor(BaseHttpHandler): вынести все адаптеры в свои отдельные файлы

loginov-dv committed:
style(TasksHandler): исправить checkstyle

loginov-dv committed:
style(SubtasksHandler): исправить checkstyle

Sla-als commented on src/ru/yandex/practicum/server/EpicsHandler.java:
```diff
@@ -0,0 +1,158 @@
+                if (epic.getID() <= 0) { // Передан epic без id - создаём новый эпик в трекере
+                    // TODO: не нужно принимать статус
+                    epic = new Epic(taskManager.nextId(), epic.getName(), epic.getDescription(),
+                            epic.getStatus());
```
Нужно убрать передачу epic.getStatus() в конструктор, использовать TaskStatus.NEW по умолчанию.

loginov-dv replied:
Исправлено, теперь для всех вновь создаваемых эпиков используется конструктор всего лишь с тремя параметрами (`id`, `name`, `description`), статус по умолчанию присваивается `NEW`

Sla-als commented on src/ru/yandex/practicum/server/BaseHttpHandler.java:
Не реализована проверка на пересечение задач в EpicsHandler, в отличие от TasksHandler и SubtasksHandler. Хотя эпик не имеет времени выполнения, лучше поддерживать консистентность логики

loginov-dv replied:
Исправлено, добавлен блок `catch` для обработки соответствующего исключения (`TaskOverlapException`)

loginov-dv committed:
feat(Epic): удалить параметр status в конструкторе для вновь создаваемых эпиков

TaskStatus для Epic - расчётный параметр, поэтому для новых эпиков по умолчанию устанавливается в NEW.

loginov-dv committed:
feat(): заменить конструктор для новых экземпляров Epic

loginov-dv committed:
feat(TaskManagerTest): заменить конструктор для новых экземпляров Epic

loginov-dv committed:
feat(InMemoryTaskManager): добавить в методы update выбрасывание исключения, если задача не найдена

loginov-dv committed:
feat(): добавить обработчики POST-запросов обработку исключения TaskNotFoundException

loginov-dv committed:
feat(EpicsHandler): добавить в обработчик POST-запросов обработку исключения TaskOverlapException
